File: lib/data/services/npm_service.dart
import 'package:dio/dio.dart';
import '../models/npm_package.dart';
import 'http_client.dart';

/// npm Registry API Service for fetching package data
class NpmService {
  final Dio _dio = HttpClient.createNpmClient();

  /// Fetch trending npm packages
  ///
  /// Uses npm registry search API with quality/popularity scoring
  /// [timeframe] - Not used by npm API, but kept for consistency
  Future<List<NpmPackage>> getTrending(String timeframe) async {
    try {
      final response = await _dio.get(
        '/-/v1/search',
        queryParameters: {
          'text': 'boost-exact:false',
          'size': 50,
          'quality': 0.8,
          'popularity': 0.9,
          'maintenance': 0.5,
        },
      );

      if (response.statusCode == 200 && response.data != null) {
        final results = response.data['objects'] as List? ?? [];
        return results.map((item) {
          final package = item['package'] as Map<String, dynamic>;
          return NpmPackage.fromJson(package);
        }).toList();
      }

      return [];
    } on DioException catch (e) {
      _handleError(e);
      return [];
    } catch (e) {
      print('Unexpected error fetching trending packages: $e');
      return [];
    }
  }

  /// Search npm packages
  ///
  /// [query] - Search query string
  /// [size] - Number of results (default: 30)
  Future<List<NpmPackage>> search({
    required String query,
    int size = 30,
  }) async {
    try {
      final response = await _dio.get(
        '/-/v1/search',
        queryParameters: {
          'text': query,
          'size': size,
        },
      );

      if (response.statusCode == 200 && response.data != null) {
        final results = response.data['objects'] as List? ?? [];
        return results.map((item) {
          final package = item['package'] as Map<String, dynamic>;
          return NpmPackage.fromJson(package);
        }).toList();
      }

      return [];
    } on DioException catch (e) {
      _handleError(e);
      return [];
    } catch (e) {
      print('Unexpected error searching packages: $e');
      return [];
    }
  }

  /// Get package details
  ///
  /// [packageName] - Full package name (e.g., 'react', '@angular/core')
  Future<NpmPackage?> getPackageDetails(String packageName) async {
    try {
      // URL encode the package name to handle scoped packages
      final encodedName = Uri.encodeComponent(packageName);

      final response = await _dio.get('/$encodedName');

      if (response.statusCode == 200 && response.data != null) {
        return NpmPackage.fromJson(response.data as Map<String, dynamic>);
      }

      return null;
    } on DioException catch (e) {
      _handleError(e);
      return null;
    } catch (e) {
      print('Unexpected error fetching package details: $e');
      return null;
    }
  }

  /// Get package downloads statistics
  ///
  /// [packageName] - Package name
  /// [period] - 'last-day', 'last-week', or 'last-month'
  Future<Map<String, dynamic>?> getDownloadStats(
    String packageName, {
    String period = 'last-week',
  }) async {
    try {
      final downloadsApiUrl = 'https://api.npmjs.org';
      final response = await Dio().get(
        '$downloadsApiUrl/downloads/point/$period/$packageName',
      );

      if (response.statusCode == 200 && response.data != null) {
        return response.data as Map<String, dynamic>;
      }

      return null;
    } on DioException catch (e) {
      print('Error fetching download stats: ${e.message}');
      return null;
    } catch (e) {
      print('Unexpected error fetching download stats: $e');
      return null;
    }
  }

  /// Handle Dio errors
  void _handleError(DioException error) {
    switch (error.type) {
      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        print('npm API Timeout error: ${error.message}');
        break;
      case DioExceptionType.badResponse:
        final statusCode = error.response?.statusCode;
        if (statusCode == 404) {
          print('npm package not found');
        } else {
          print('npm API error: $statusCode - ${error.response?.data}');
        }
        break;
      case DioExceptionType.cancel:
        print('npm API request cancelled');
        break;
      default:
        print('npm API connection error: ${error.message}');
    }
  }
}

File: packages/app/src/api/npm/npmService.ts
/**
 * npm API Service
 * Business logic for npm Registry API interactions
 * Ported from Flutter app: lib/data/services/npm_service.dart
 */

import { npmClient } from './npmClient';
import { NpmPackage } from '../../models/NpmPackage';
import { TimeframeType } from '../github/githubClient';

export interface NpmSearchOptions {
  quality?: number; // 0-1
  popularity?: number; // 0-1
  maintenance?: number; // 0-1
  perPage?: number;
}

class NpmService {
  /**
   * Get trending npm packages
   * Note: npm doesn't have a direct "trending" endpoint, so we use search
   * with quality/popularity weights to simulate trending
   */
  async getTrending(
    timeframe: TimeframeType = 'daily',
    options: NpmSearchOptions = {},
  ): Promise<NpmPackage[]> {
    try {
      const {
        quality = 0.8,
        popularity = 0.9,
        maintenance = 0.5,
        perPage = 50,
      } = options;

      // Use npm search API with popularity/quality weights
      const response = await npmClient.getInstance().get('/-/v1/search', {
        params: {
          text: 'boost-exact:false', // Get broader results
          size: perPage,
          quality,
          popularity,
          maintenance,
        },
      });

      // Parse and return packages
      const packages = response.data.objects.map((obj: any) => {
        const pkg = NpmPackage.fromJSON(obj.package);
        pkg.trendingPeriod = timeframe;
        return pkg;
      });

      return packages;
    } catch (error) {
      console.error('Error fetching trending packages:', error);
      throw new Error('Failed to fetch trending packages');
    }
  }

  /**
   * Search npm packages
   */
  async search(
    query: string,
    options: NpmSearchOptions = {},
  ): Promise<NpmPackage[]> {
    try {
      const {
        quality = 0.5,
        popularity = 0.5,
        maintenance = 0.5,
        perPage = 50,
      } = options;

      const response = await npmClient.getInstance().get('/-/v1/search', {
        params: {
          text: query,
          size: perPage,
          quality,
          popularity,
          maintenance,
        },
      });

      return response.data.objects.map((obj: any) =>
        NpmPackage.fromJSON(obj.package),
      );
    } catch (error) {
      console.error('Error searching packages:', error);
      throw new Error('Failed to search packages');
    }
  }

  /**
   * Get package details
   */
  async getPackage(packageName: string): Promise<NpmPackage> {
    try {
      const response = await npmClient.getInstance().get(`/${packageName}`);
      return NpmPackage.fromJSON(response.data);
    } catch (error) {
      console.error('Error fetching package details:', error);
      throw new Error('Failed to fetch package details');
    }
  }

  /**
   * Get package download stats
   */
  async getDownloadStats(
    packageName: string,
    period: 'last-day' | 'last-week' | 'last-month' = 'last-week',
  ) {
    try {
      const response = await npmClient
        .getInstance()
        .get(`https://api.npmjs.org/downloads/point/${period}/${packageName}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching download stats:', error);
      return null;
    }
  }

  /**
   * Get package versions
   */
  async getVersions(packageName: string): Promise<string[]> {
    try {
      const response = await npmClient.getInstance().get(`/${packageName}`);
      const versions = response.data.versions;
      return Object.keys(versions || {});
    } catch (error) {
      console.error('Error fetching versions:', error);
      return [];
    }
  }

  /**
   * Search packages by keyword
   */
  async searchByKeyword(
    keyword: string,
    options: NpmSearchOptions = {},
  ): Promise<NpmPackage[]> {
    try {
      const { perPage = 50 } = options;

      const response = await npmClient.getInstance().get('/-/v1/search', {
        params: {
          text: `keywords:${keyword}`,
          size: perPage,
        },
      });

      return response.data.objects.map((obj: any) =>
        NpmPackage.fromJSON(obj.package),
      );
    } catch (error) {
      console.error('Error searching by keyword:', error);
      return [];
    }
  }

  /**
   * Get recently updated packages
   */
  async getRecentlyUpdated(
    limit: number = 20
  ): Promise<NpmPackage[]> {
    try {
      const response = await npmClient.getInstance().get('/-/v1/search', {
        params: {
          text: 'boost-exact:false',
          size: limit,
          from: 0,
          quality: 0.5,
          popularity: 0.8,
          maintenance: 0.9, // Prioritize well-maintained packages
        },
      });

      if (!response.data || !response.data.objects) {
        return [];
      }

      // Sort by modified date (most recent first)
      const sortedObjects = response.data.objects.sort((a: any, b: any) => {
        const dateA = new Date(a.package.date).getTime();
        const dateB = new Date(b.package.date).getTime();
        return dateB - dateA;
      });

      return sortedObjects
        .slice(0, limit)
        .map((obj: any) => NpmPackage.fromJSON(obj.package));
    } catch (error) {
      console.error('Error fetching recently updated packages:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const npmService = new NpmService();
export default npmService;
